File: ios/Runner/MethodChannel/QDComplianceMethodChannel.swift
//
//  QDComplianceMethodChannel.swift
//  Runner
//
//  Created by saminos on 04/03/22.
//

import Foundation
import QcmpSDK

class QDComplianceMethodChannel: FlutterMethodChannel {
    private static let compliance = Compliance.shared
    private static let channel = "io.quadrant.compliance"
    
    @discardableResult
    static func setup(_ binaryMessenger: FlutterBinaryMessenger, integrationKey: String) -> QDComplianceMethodChannel {
        
        let methodChannel = QDComplianceMethodChannel(name: Self.channel, binaryMessenger: binaryMessenger)
        compliance.setup(integrationKey)
        methodChannel.setup()
        return methodChannel
    }
    
    private func setup() {
        setMethodCallHandler(Self.handler(call:result:))
    }
    
    private static func handler(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "openConsentFormIfNeeded":
            compliance.openConsentFormIfNeeded { (consentResult) in
                switch consentResult {
                case .success(let choosenConsent):
                    result(choosenConsent.rawValue)
                case .failure(let error):
                    result(FlutterError(code: "", message: error.errorDescription, details: nil))
                }
            }
            
        case "openConsentForm":
            guard let arguments = call.arguments as? [String: Any],
                  let complianceTypeRaw = arguments["complianceType"] as? Int,
                  let type = ComplianceType(rawValue: complianceTypeRaw)
            else { return }
            
            compliance.openConsentForm(complianceType: type) { consentResult in
                switch consentResult {
                case .success(let choosenConsent):
                    result(choosenConsent.rawValue)
                case .failure(let error):
                    result(FlutterError(code: "", message: error.errorDescription, details: nil))
                }
            }
            case "optOut":
            compliance.requestCompliance(requestType: .REQUEST_OPTOUT) { requestResult in
                switch requestResult {
                case .success:
                    result(true)
                case .failure(let error):
                    result(FlutterError(code: "", message: error.errorDescription, details: nil))
                }
            }
            case "doNotSell":
            compliance.requestCompliance(requestType: .REQUEST_DO_NOT_SELL) { requestResult in
                switch requestResult {
                case .success:
                    result(true)
                case .failure(let error):
                    result(FlutterError(code: "", message: error.errorDescription, details: nil))
                }
            }
            case "deleteMyData":
            compliance.requestCompliance(requestType: .REQUEST_DELETE_DATA) { requestResult in
                switch requestResult {
                case .success:
                    result(true)
                case .failure(let error):
                    result(FlutterError(code: "", message: error.errorDescription, details: nil))
                }
            }
            case "requestMyData":
            compliance.requestCompliance(requestType: .REQUEST_ACCESS_DATA) { requestResult in
                switch requestResult {
                case .success:
                    result(true)
                case .failure(let error):
                    result(FlutterError(code: "", message: error.errorDescription, details: nil))
                }
            }
            
        default: break;
        }
    }
}

File: android/app/src/main/kotlin/com/example/flutter_compliance/QDComplianceMethodChannel.kt
package com.example.compliance

import android.content.Context
import androidx.fragment.app.FragmentManager
import com.quadrant.sdk.compliance.core.Compliance
import com.quadrant.sdk.compliance.util.ConsentResult
import com.quadrant.sdk.compliance.util.PermissionUtil
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel

class QDComplianceMethodChannel(context: Context, fragmentManager: FragmentManager, flutterEngine: FlutterEngine, integrationKey: String) {

    private val CHANNEL = "io.quadrant.compliance"
    private val REQUEST_OPTOUT = 0
    private val REQUEST_DO_NOT_SELL = 1
    private val REQUEST_DELETE_DATA = 2
    private val REQUEST_ACCESS_DATA = 3

    var compliance: Compliance = Compliance.getInstance(integrationKey)
    init {

        MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL).setMethodCallHandler {
            call, result ->

            when (call.method) {
                "openConsentFormIfNeeded" -> openConsentFormIfNeeded(context, fragmentManager, result)
                "openConsentForm" -> openConsentForm(fragmentManager, call.argument<Int>("complianceType"), result)
                "optOut" -> optOut(context, fragmentManager,result)
                "doNotSell" -> doNotSell(context, fragmentManager,result)
                "deleteMyData" -> deleteMyData(context, fragmentManager,result)
                "requestMyData" -> requestMyData(context, fragmentManager,result)
            }
        }
    }

    private fun openConsentFormIfNeeded(context: Context, fragmentManager: FragmentManager, result: MethodChannel.Result) {
        if (!PermissionUtil.accessLocation(context)) return
        compliance.openConsentFormIfNeeded(context, fragmentManager, object: Compliance.ConsentCallback {
            override fun onSuccess(consentResult: ConsentResult?) {
                result.success(consentResult.toString())
            }

            override fun onError(errorMessage: String?) {
               result.error("", errorMessage,null)
            }

        })
    }

    private fun openConsentForm( fragmentManager: FragmentManager, complianceType: Int?, result: MethodChannel.Result) {
        complianceType?.let {
            compliance.openConsentForm(fragmentManager, complianceType, object : Compliance.ConsentCallback {
                override fun onSuccess(consentResult: ConsentResult?) {
                    result.success(consentResult.toString())
                }

                override fun onError(errorMessage: String?) {
                    result.error("", errorMessage,null)
                }
            })
        }
    }

    private fun optOut(context: Context, fragmentManager: FragmentManager, result: MethodChannel.Result) {
        compliance.requestCompliance(context, fragmentManager, REQUEST_OPTOUT, object: Compliance.ResultCallback {
            override fun onSuccess(p0: String?) {
                result.success(p0)
            }

            override fun onError(errorMessage: String?) {
                result.error("", errorMessage,null)
            }
        })
    }

    private fun doNotSell(context: Context, fragmentManager: FragmentManager, result: MethodChannel.Result) {
        compliance.requestCompliance(context, fragmentManager, REQUEST_DO_NOT_SELL, object: Compliance.ResultCallback {
            override fun onSuccess(p0: String?) {
                result.success(p0)
            }

            override fun onError(errorMessage: String?) {
                result.error("", errorMessage,null)
            }
        })
    }

    private fun deleteMyData(context: Context, fragmentManager: FragmentManager, result: MethodChannel.Result) {
        compliance.requestCompliance(context, fragmentManager, REQUEST_DELETE_DATA, object: Compliance.ResultCallback {
            override fun onSuccess(p0: String?) {
                result.success(p0)
            }

            override fun onError(errorMessage: String?) {
                result.error("", errorMessage,null)
            }
        })
    }

    private fun requestMyData(context: Context, fragmentManager: FragmentManager, result: MethodChannel.Result) {
        compliance.requestCompliance(context, fragmentManager, REQUEST_ACCESS_DATA, object: Compliance.ResultCallback {
            override fun onSuccess(p0: String?) {
                result.success(p0)
            }

            override fun onError(errorMessage: String?) {
                result.error("", errorMessage,null)
            }
        })
    }
}

